drv/lpc55-swd/src/armv7debug.rs

```diff
@@ -0,0 +1,189 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//
+// For the STM32H7xx, see ARMv7-M Architecture Reference Manual
+// Part 3 Debug Arch.
+// https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture?lang=en
+
+use bitflags::bitflags;
+
+pub trait DpAddressable {
+    // Address for accessing a DP Register
+    fn addr() -> u32;
+}
+
+// For keeping track of unwinding debug actions
+bitflags! {
+    #[derive(PartialEq, Eq, Copy, Clone)]
+    pub struct Undo: u8 {
+        // Need self.swd_finish()
+        const SWD = 1 << 0;
+        // Need self.sp_reset_leave(true)
+        const RESET = 1 << 1;
+        // Need DEMCR = 0
+        const VC_CORERESET = 1 << 2;
+        // Need to clear debug enable.
+        const DEBUGEN = 1 << 3;
+    }
+}
+
+// RW   0x00000000    Debug Halting Control and Status Register
+// Some DHCSR bits have different read vs. write meanings
+// Specifically, the MAGIC value enables writing other control bits
+// but some of those bits are status bits when read.
+bitflags! {
+    #[derive(PartialEq, Eq, Copy, Clone)]
+    pub struct Dhcsr: u32 {
+        // At least one reset since last DHCSR read. clear on read.
+        const S_RESET_ST = 1 << 25;
+        const S_RETIRE_ST = 1 << 24;
+        const S_LOCKUP = 1 << 19;
+        const S_SLEEP = 1 << 18;
+        const S_HALT = 1 << 17;
+        const S_REGRDY = 1 << 16;
+
+        // Magic number allows setting C_* bits.
+        const DBGKEY = 0xA05F << 16;
+
+        const C_SNAPSTALL = 1 << 5;
+        const C_MASKINTS = 1 << 3;
+        const C_STEP = 1 << 2;
+        const C_HALT = 1 << 1;
+        const C_DEBUGEN = 1 << 0;
+        const _ = !0;
+    }
+}
+
+impl From<u32> for Dhcsr {
+    fn from(v: u32) -> Self {
+        Self::from_bits_retain(v)
+    }
+}
+
+impl DpAddressable for Dhcsr {
+    fn addr() -> u32 {
+        Self::ADDRESS
+    }
+}
+
+impl Dhcsr {
+    const ADDRESS: u32 = 0xE000EDF0;
+    pub fn halt() -> Self {
+        Self::DBGKEY | Self::C_HALT | Self::C_DEBUGEN
+    }
+    pub fn resume() -> Self {
+        Self::DBGKEY | Self::C_DEBUGEN
+    }
+    pub fn end_debug() -> Self {
+        Self::DBGKEY
+    }
+    pub fn is_halted(self) -> bool {
+        self & Self::S_HALT == Self::S_HALT
+    }
+    pub fn is_regrdy(self) -> bool {
+        self & Self::S_REGRDY == Self::S_REGRDY
+    }
+    pub fn _is_lockup(self) -> bool {
+        self & Self::S_LOCKUP == Self::S_LOCKUP
+    }
+}
+
+// Debug Core Register Selector Register
```

Comment:
I like the new `DpAddressable` + `bitflags!` design, so it's awkward that these two registers continue to use hard-coded values...